src/privatebin.rs

```diff
@@ -165,6 +174,71 @@ impl Paste {
         let key = bs58::decode(bs58_key).into_vec()?;
         crate::crypto::decrypt_with_password(self, &key, password)
     }
+
+    /// Returns a mapping: comment.id -> decrypted_comment
+    pub fn decrypt_comments(&self, bs58_key: &str) -> PbResult<DecryptedCommentsMap> {
+        self.decrypt_comments_with_password(bs58_key, "")
+    }
+
+    /// Returns a mapping: comment.id -> decrypted_comment
+    pub fn decrypt_comments_with_password(
+        &self,
+        bs58_key: &str,
+        password: &str,
+    ) -> PbResult<DecryptedCommentsMap> {
+        let mut decrypted_comments = HashMap::new();
+        if let Some(comments) = &self.comments {
+            for comment in comments {
+                let id = comment.id.clone();
+                decrypted_comments.insert(id, comment.decrypt_with_password(bs58_key, password)?);
+            }
+        };
+        Ok(decrypted_comments)
+    }
+
+    /// Returns a mapping: comment.id -> [children comment.id]
+    pub fn comments_adjacency_map(&self) -> PbResult<CommentsAdjacencyMap> {
+        let mut comment_adjacency: CommentsAdjacencyMap = HashMap::new();
+        if let Some(comments) = &self.comments {
+            for c in comments {
+                let id = c.id.clone();
+                let parentid = c.parentid.clone();
+                comment_adjacency.entry(parentid).or_default().push(id);
+            }
+        }
+        Ok(comment_adjacency)
+    }
+
+    /// Returns formatted json trees (forest) of decrypted comments
+    pub fn comments_formatted_json_trees(
+        &self,
+        decrypted_comments: &DecryptedCommentsMap,
+        comment_adjacency: &CommentsAdjacencyMap,
+    ) -> PbResult<String> {
+        fn format_comments_below_id(
+            id: &str,
+            decrypted_comments: &DecryptedCommentsMap,
+            comment_adjacency: &CommentsAdjacencyMap,
+        ) -> serde_json::Value {
+            let formatted_children: Vec<serde_json::Value> = comment_adjacency
+                .get(id)
+                .unwrap_or(&Vec::new())
+                .iter()
+                .map(|child_id| {
+                    format_comments_below_id(child_id, decrypted_comments, comment_adjacency)
+                })
+                .collect();
+            json!({
+                id: {
+                    "comment": decrypted_comments.get(id).unwrap_or(&DecryptedComment::default()).comment,
+                    "nickname": decrypted_comments.get(id).unwrap_or(&DecryptedComment::default()).nickname,
+                    "replies": formatted_children
+                }
+            })
+        }
+        let top_level = format_comments_below_id(&self.id, decrypted_comments, comment_adjacency);
+        serde_json::to_string_pretty(&top_level[&self.id]["replies"]).map_err(|e| e.into())
```

Comment:
Would it make more sense to just return a serde value here? Since on the caller side of the function we immediately parse the string back into a json value

Reply:
This function will have other users (for example uniffi export), where the simpler `String` output would be preferred.   `serde_json::Value` being an external type would limit portability from a library point of view.

Reply:
Hm, it appears github swallowed my comment.

I suspected as such. I am currently looking into uniffi in a separate branch, and once I get the hang of it I believe it should be possible to create a custom type for it which translates between string and serde::value. 

But for for now, I would proceed to merge this as is and look into that once I warmed up with uniffi